File: 2025/684. Redundant Connection/Solution.java
class UnionFind {
    int[] parents;
    public UnionFind(int n) {
        parents = new int[n];
        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }
    }
    public int find(int x) {
        if (parents[x] != x) {
            parents[x] = find(parents[x]);
        }
        return parents[x];
    }
    public boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX == rootY) {
            return true;
        }
        parents[rootX] = rootY;
        return false;
    }
}

class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        UnionFind uf = new UnionFind(edges.length);
        for (int[] e : edges) {
            if (uf.union(e[0] - 1, e[1] - 1)) {
                return e;
            }
        }
        return new int[0];
    }
}

File: 2025/684. Redundant Connection/solution.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
    
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return True
        self.parent[root_x] = root_y
        return False

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        unionfind = UnionFind(len(edges))
        for a, b in edges:
            if unionfind.union(a - 1, b - 1):
                return [a,b]
        return []